compiler/src/dotty/tools/dotc/core/Constants.scala

```diff
@@ -161,25 +163,29 @@ object Constants {
           }
         case pt => pt
       }
-      val target = classBound(pt).typeSymbol
-      if (target == tpe.typeSymbol)
-        this
-      else if ((target == defn.ByteClass) && isByteRange)
-        Constant(byteValue)
-      else if (target == defn.ShortClass && isShortRange)
-        Constant(shortValue)
-      else if (target == defn.CharClass && isCharRange)
-        Constant(charValue)
-      else if (target == defn.IntClass && isIntRange)
-        Constant(intValue)
-      else if (target == defn.LongClass && isLongRange)
-        Constant(longValue)
-      else if (target == defn.FloatClass && isFloatRange)
-        Constant(floatValue)
-      else if (target == defn.DoubleClass && isNumeric)
-        Constant(doubleValue)
-      else
-        null
+      pt match
+        case ConstantType(value) if value == this => this
```

Comment:
Idle thought, what would it look like with guard terminated by `then`? `case P if b then body`

Reply:
Too late for that...